how-it-works/1_protocol/03_message-routing.subpage.md

```diff
@@ -7,8 +7,128 @@ slug: message-routing
+  * Canister metadata (module hashes, *certified variables*)
+* Items certified per checkpoint:
+  * All canister Wasm bytecode
+  * All canister state (heap and stable memory)
 
+The message routing layer performs a certification of parts of the replicated state after each round (the first three items above) and certification of the full replicated state (including also the remaining two items above) at every checkpoint, which is typically every multiple-hundred rounds.
+Certification is always done using BLS threshold signatures computed collectively by the subnet, thus certifications are computed by the subnet as a whole in a decentralized manner.
+
+### Per-Round Certification
+
+At the end of a round, i.e., when all messages have been executed or the cycles limit for the round has been reached (to ensure rounds cannot take arbitrarily long), the message routing layer performs a certification of parts of the replicated state.
+A BLS threshold signature is computed to certify the part of the state tree containing
+* Responses to ingress messages,
+* Xnet messages to be sent to other subnets, and
+* Canister metadata (module hashes, *certified variables*).
+
+The responses to ingress messages are referred to as ingress history.
+The certified responses can be read and validated against the subnet's public key by users as the response to their ingress messages.
+This way of validating responses to state-changing messages to a blockchain is extremely powerful when compared to other approaches seen in the field like reading the response from a transaction log.
+
+This per-round state certification ensures that any item of data relevant for interactions of users and subnets and between different subnets on the Internet Computer is authenticated.
+This particularly enables secure and verifiable inter-subnet communication, a crucial feature of the Internet Computer as well as an enabler of its scalability.
+
+### Per-Checkpoint Certification
+
+Wasm code changed through canister updates and written-to ("dirty") memory pages of canisters are not certified in every round, but flagged as modified by the execution layer whenever they are written, using approaches known from operating system kernels.
+At every checkpointing interval, that is, every multiple hundred rounds, which corresponds to around 10 minutes, the subnet computes a so-called checkpoint at which the state of a subnet is certified.
+The state certification is done incrementally by incorporating the content of all pages changed since the last checkpoint certification into the Merkle state tree and propagating the changes up the tree.
+Finally, the root of the Merkle tree and additional metadata is signed with a BLS threshold signature by the subnet, thereby certifying every leaf of the Merkle tree.
+The signed result is called a *catch-up package* as it can be used by nodes to efficiently catch up to the point in time when the checkpoint was made.
```

Comment:
```suggestion
The signed result is called a *catch-up package* as it can be used by nodes to efficiently catch up to the point in time when the checkpoint was made. (Note that a catch-up package also contains other things required to resume, which are omitted here for the sake of simplicity).
```